### ISSUE TITLE
ERL-421: ODBC truncates data for LONGTEXT and BLOB

### ISSUE THREAD

**Issue by OTP-Maintainer** [erlang/otp#3321]
Original reporter: `lea`
Affected version: `OTP-19.1.1`
Component: `odbc`
Migrated from: https://bugs.erlang.org/browse/ERL-421

---

```
Tested with MySQL 5.6.36 on CentOS 6.9 64-bit and rpm esl-erlang-19.3-1.x86_64.

If you have either a LONGTEXT or BLOB column, content exceeding 8001 bytes will be truncated when reading from the database.

+To reproduce:+
Create a table customerField in a database named tull and write content of e.g. 10000 characters in a LONGTEXT column. Then, from the erlang shell:

odbc:start().                                                                                           
Connect="DSN=sqlconnection;UID=root;PWD=;MULTI_STATEMENTS=1;BIG_PACKETS=1;charset=utf8mb4;database=tull".
{ok,Ref} = odbc:connect(Connect,[]).                                                                    
odbc:sql_query(Ref,"select value from sessionField ").                                                  

+Reason for problem:+
Looking at ../lib/odbc/c_src/odbcserver.c and odbcserver.h:

In function encode_column_name_list() there is an if clause:
	if(sql_type == SQL_LONGVARCHAR || sql_type == SQL_LONGVARBINARY || sql_type == SQL_WLONGVARCHAR)
	    size = MAXCOLSIZE;

where MAXCLOSIZE is defined to 8001.

+Suggested solution:+
Some kind of dynamic reading using SQLGetData.
```

**Comment by winks** [comment]
I spent some time debugging this today (on today's master), and here are some findings:

When connecting to Postgres 14.1, and inspecting some values in the aforementioned file `lib/odbc/c_src/odbcserver.c`:

```
# create table barb (bfield bytea);
# select bfield from barb;
size 0, sql_type -4 (SQL_LONGVARBINARY)

# create table bart (tfield text);
# select tfield from bart;
size 8190, sql_type -1 (SQL_LONGVARCHAR)
```

So the above mentioned patch by jan.chochol seems to be fine for Postgres.

When connecting to Oracle, unfortunately it looks quite different:

```
# create table barb (bfield blob); insert ... (a BLOB of length 8765)
# select bfield from barb;
size 2147483647, sql_type -4 (SQL_LONGVARBINARY)

# create table bart (bfield clob); insert ... (a CLOB of length 16005)
# select tfield from bart;
size 2147483647, sql_type -1 (SQL_LONGVARCHAR)
```

Playing around and setting `size = MAXCOLSIZE*3` seems to give me the data but that's just another arbitrary limit. (But the length() of the returned data is correct, as long as it's below the MAXCOLSIZE*3)

Maybe I can also try mysql, but I haven't done it so far.

I also read a few other codebases, so for reference here are calls to `SQLDescribeCol` in

 * pyodbc: https://github.com/mkleehammer/pyodbc/blob/24e0ccadaec091ce6b1b81e3197a345b895773fc/src/cursor.cpp#L167
 * pyodbc: https://github.com/mkleehammer/pyodbc/blob/24e0ccadaec091ce6b1b81e3197a345b895773fc/src/cursor.cpp#L487
 * PHP: https://github.com/php/php-src/blob/19061e3b4354e384ffb70d401afa9004e2327753/ext/pdo_odbc/odbc_stmt.c#L561

In PyODBC it just seems to work, but I couldn't yet figure out what's different.

In PHP based on the type of the field it's set to retrieve data later, not unlike `retrive_binary_data` (sic!) here, but to me it seems as if `(columns(state)[i]).type.c` is never set to `SQL_BINARY` and thus `retrive_binary_data` is not invoked.

TLDR: I think the proposed patch is not a full fix, at least not when Oracle is involved.

**Comment by garazdawi** [comment]
Have you also checked with #4822? It may fix some more things, but I don't think it fixes all things.
